plugin/plugin.go

```diff
@@ -141,11 +132,186 @@ func (p *AzurePlugin) OnUpdateSet(_ context.Context, req *pb.OnUpdateSetRequest)
+
+			if err := p.processStandardVMs(ctx, vmResources, clients, vmNetworkMap); err != nil {
+				return fmt.Errorf("processing VMs: %w", err)
+			}
+
+			mu.Lock()
+			defer mu.Unlock()
+			for k, v := range vmNetworkMap {
+				vmToNetworkMap[k] = v
+			}
+
+			fmt.Printf("Processed %d VMs in %v\n", len(vmResources), time.Since(vmStart))
+			return nil
+		})
+	}
+
+	// Process VMSS in parallel if they exist
+	if len(vmssResources) > 0 {
+		g.Go(func() error {
+			vmssStart := time.Now()
+			vmssNetworkMap := make(map[string]networkInfo)
+
+			if err := p.processVMScaleSets(ctx, vmssResources, clients, vmssNetworkMap); err != nil {
+				return fmt.Errorf("processing VMSS: %w", err)
+			}
+
+			mu.Lock()
+			defer mu.Unlock()
+			for k, v := range vmssNetworkMap {
+				vmToNetworkMap[k] = v
+			}
+
+			fmt.Printf("Processed %d VMSS in %v\n", len(vmssResources), time.Since(vmssStart))
+			return nil
+		})
+	}
+
+	// Wait for all goroutines to complete and check for errors
+	if err := g.Wait(); err != nil {
+		return nil, err
+	}
+	fmt.Printf("Total processing time: %v\n", time.Since(processingStart))
+
+	responseStart := time.Now()
+	response := p.buildHostsResponse(vmToNetworkMap, resourceToSetMap)
+	fmt.Printf("Built response in %v (Total hosts: %d)\n",
+		time.Since(responseStart),
+		len(vmToNetworkMap))
+
+	return response, nil
+}
+
+func rotateCredFromCallback(ctx context.Context, catalog *hostcatalogs.HostCatalog) (*pb.HostCatalogPersisted, error) {
+	authzInfo, err := getAuthorizationInfo(catalog)
+	if err != nil {
+		return nil, fmt.Errorf("error getting auth config when creating catalog: %w", err)
+	}
+	newPass, err := rotateCredential(ctx, authzInfo)
+	if err != nil {
+		return nil, fmt.Errorf("error rotating credentials when creating catalog: %w", err)
+	}
+	if newPass == nil {
+		return nil, errors.New("new credential back from rotate is nil")
+	}
+	newSecrets, err := structpb.NewStruct(map[string]interface{}{
+		constSecretId:             *newPass.KeyId,
+		constSecretValue:          *newPass.SecretText,
+		constCredsLastRotatedTime: time.Now().Format(time.RFC3339Nano),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error formatting new credential data as struct")
+	}
+	return &pb.HostCatalogPersisted{
+		Secrets: newSecrets,
+	}, err
+}
+
+// -----------------------------------------------------------------------------
+// Response Building Functions
+// -----------------------------------------------------------------------------
+func (p *AzurePlugin) buildHostsResponse(vmToNetworkMap map[string]networkInfo,
+	resourceToSetMap map[string][]string) *pb.ListHostsResponse {
+
+	ret := &pb.ListHostsResponse{}
+
+	for resourceID, networkInfo := range vmToNetworkMap {
+		host := p.createHostFromResource(resourceID, networkInfo, resourceToSetMap)
+		if host != nil {
+			ret.Hosts = append(ret.Hosts, host)
+		}
+	}
+
+	return ret
+}
+
+func (p *AzurePlugin) createHostFromResource(resourceID string, networkInfo networkInfo,
```

Comment:
Doesn't look like these need to be part of the AzurePlugin methods?
```suggestion
func buildHostsResponse(vmToNetworkMap map[string]networkInfo,
	resourceToSetMap map[string][]string) *pb.ListHostsResponse {

	ret := &pb.ListHostsResponse{}

	for resourceID, networkInfo := range vmToNetworkMap {
		host := createHostFromResource(resourceID, networkInfo, resourceToSetMap)
		if host != nil {
			ret.Hosts = append(ret.Hosts, host)
		}
	}

	return ret
}

func createHostFromResource(resourceID string, networkInfo networkInfo,
```